packages/smart-wallet/src/smartWallet.js

```diff
@@ -0,0 +1,307 @@
+   *
+   * @type {MapStore<Brand, ERef<Purse<'set'>>>}
+   */
+  const brandPurses = makeScalarBigMapStore('brand purses', { durable: true });
+
+  /** @type { (desc: Omit<BrandDescriptor, 'displayInfo'>) => Promise<void>} */
+  const addBrand = async desc => {
+    console.log('addBrand', desc);
+    // assert haven't received this issuer before.
+    const descriptorsHas = brandDescriptors.has(desc.brand);
+    const pursesHas = brandPurses.has(desc.brand);
+    assert(
+      !(descriptorsHas && pursesHas),
+      'repeated brand from bank asset subscription',
+    );
+    assert(
+      !(descriptorsHas || pursesHas),
+      'corrupted state; one store has brand already',
+    );
+
+    const [purse, displayInfo] = await Promise.all([
+      E(desc.issuer).makeEmptyPurse(),
+      E(desc.brand).getDisplayInfo(),
+    ]);
+
+    // save all five of these in a collection (indexed by brand?) so that when
+    // it's time to take an offer description you know where to get the
+    // relevant purse. when it's time to make an offer, you know how to make
+    // payments. REMEMBER when doing that, need to handle every exception to
+    // put the money back in the purse if anything fails.
+    const descriptor = { ...desc, displayInfo };
+    brandDescriptors.init(desc.brand, descriptor);
+    brandPurses.init(desc.brand, purse);
+
+    // publish purse's balance and changes
+    E.when(
+      purse.getCurrentAmount(),
+      balance => updateBalance(purse, balance, 'init'),
+      err => console.error('initial purse balance publish failed', err),
+    );
+    observeNotifier(E(purse).getCurrentAmountNotifier(), {
+      updateState(balance) {
+        updateBalance(purse, balance);
+      },
+      fail(reason) {
+        console.error(`failed updateState observer`, reason);
+      },
+    });
+
+    updatePublishKit.publisher.publish({ updated: 'brand', descriptor });
+  };
+
+  // Ensure a purse for each issuer
+  addBrand({
+    brand: invitationBrand,
+    issuer: invitationIssuer,
+    petname: 'invitations',
+  });
+  // watch the bank for new issuers to make purses out of
+  observeIteration(E(bank).getAssetSubscription(), {
+    async updateState(desc) {
+      await addBrand({
+        brand: desc.brand,
+        issuer: desc.issuer,
+        petname: desc.proposedName,
+      });
+    },
+  });
+  // #endregion
+
+  /**
+   * Similar to {DepositFacet} but async because it has to look up the purse.
+   */
+  const depositFacet = Far('smart wallet deposit facet', {
+    /**
+     * Put the assets from the payment into the appropriate purse
+     *
+     * @param {ERef<Payment>} paymentE
+     * @param {Brand} paymentBrand must match the payment's brand. passed in to save lookup.
+     * @returns {Promise<Amount>}
+     * @throws if the purse doesn't exist
+     * ??? PRODUCT: should it instead wait for an issuer/purse to put it in?
+     */
+    receive: (paymentE, paymentBrand) => {
+      const purse = brandPurses.get(paymentBrand);
```

Comment:
Integration testing suggests that `paymentBrand` is optional.

```suggestion
      const brand = await (paymentBrand || E(paymentE).getAllegedBrand());
      const purse = brandPurses.get(brand);
```